39 implement tests for nabla indexer

TorstenStueber opened the pull request:
This now ballooned into one big PR that contains many different changes implemented over months.

## Updated deployment instructions
We deployed a number of various Nabla instances in the past: different instances on Pendulum (used for Vortex and public deployments) and instances on Foucoco (used for the testnet campaign and for different validations and testing scenarios).

There was only one definition of the deployment scripts and when changing to a new deployment I had to adapt all files locally. This had a couple of issues:

* no record of historic deployment definitions
* editing the files was error prone as names had to be changed consistently across multiple files
* I had a lot of inconsistent local changes to this repository that never ended up as a proper PR – this is one of the reasons why this now such a big PR

Instead, I now changed the deployment scripts to use declarative deployments (see the folder `nabla/deployments`). That way editing is less error prone and we can easily keep a history of deployments.

For that reason there is no need for separate `nabla` and `nabla-pendulum` projects anymore. There is now just a `pendulum` project.

## Updated Tests
Closes #39 
Closes #44 
Closes https://github.com/pendulum-chain/wasm-deploy/issues/43 (fixed by [this](https://github.com/pendulum-chain/wasm-deploy/pull/40/commits/3ca21b90bfbeba48ff47ed509c8eb4a8269bf744))

This PR adds a new test suite to test the Nabla indexer (called `nabla-indexer`).

These tests are implemented for this indexer PR: https://github.com/pendulum-chain/pendulum-squids/pull/52.

The main file is `nabla-indexer/test/indexer_tests.ts`.

Additionally this ticket also updates the version of the Nabla contracts that had some minor changes in the mean time and some new tests have been implemented.

This is temporary work. The intention is to move these tests to our subsquid indexer repository and to move the Nabla tests to the Nabla contracts repository. However, this requires that wasm-deploy becomes a standalone tool that can be used in other npm projects (see https://github.com/pendulum-chain/wasm-deploy/issues/10).

Additionally, this PR adds some small new features to wasm-deploy that were necessary to implement this test suite:
- overwrite message names in the generated metadata to address (see https://github.com/hyperledger/solang/issues/1623)
- overwrite empty message argument names (otherwise Subsquid will generate improper code)
- allow to execute arbitrary blockchain root transactions in tests
- allow to specify an arbitrary commit hash to checkout for contract repositories

## Other Changes
* update all dependencies, particularly api-solang
* make `tsx` a dev dependency instead of a normal dependency
* add `lint:ts` script

TorstenStueber committed:
Create Nabla indexer tests

Create nabla-indexer project

Remove unnecessary contracts

Allow to override metadata message names

Allow to override argument names of messages in metadata

Include new tests from Nabla main repo

Update README

Use different token decimals for Nabla testnet deployment

Update names in deployment script

Add deployment for five swap pools

Adapt tests for new changes

Implement test from Alpha Testnet

Add tests for swap pool quote functions

Implement tests to change treasury

Add missing calls to Nabla deployment

Remove `api-solang` from workspaces
This is in order to use the deployed npm version with types available

Replace ts-node with tsx
ts-node was not running due to module vs commonjs issues

Remove `--ignore-engines`
because it's deprecated and the pull command throws an error

Change `init` for 'nablaCurve' repo to "npm"

Set polkadot dependencies to fixed version
There was a breaking change in a minor version bump so we set these to specific versions
see https://github.com/polkadot-js/api/commit/ca4802326c07bffc48a3b4dad7ff88b8dca3098c#diff-6c3c68a16db206b507cabe5612726492ebaf4c9f3d4717e10d2a75b3d6dcd4f1R172

Rename commands in package.json

Add info to README.md

Add fixes to make nabla indexer tests successful again

Change token names and symbols in testnet deployment

Create debug snapshot

Simplify Nabla project

TorstenStueber commented on nabla/test/swapPool.ts:
```diff
@@ -63,6 +63,44 @@ export default async function (environment: TestSuiteEnvironment) {
       await genericTestOnlyOwnerCanSetPoolCap(pool, environment);
     },
 
+    async test_maxCoverageRatioForSwapIn_CheckDefaultValue() {
```
These are additional tests from this PR in the Nabla repo: https://github.com/0xamberhq/contracts/pull/76/files#diff-ef61283d9b153c522945399fe4fa316c48029036ca90a2648efa5d5794d5c5ee

gianfra-t commented on nabla-indexer/test/indexer_tests.ts:
```diff
@@ -0,0 +1,1067 @@
+    vm.startPrank(BOB);
+    await assets[0].approve(address(backstop), MAX_UINT256);
+    for (let i = 1; i <= noOfSwapPools; i++) {
+      await assets[i].approve(address(router), MAX_UINT256);
```
Is it also required for the depositor to give approval for the router contract?

TorstenStueber replied:
Indeed that should not be required for a deposit. It is only required for a token swap. I think I wanted to do this here preemptively as I possibly planned to use it in one of the tests (i.e., that BOB would execute a swap). I didn't check whether it's still required. Should not harm to leave it in.

gianfra-t replied:
Ah makes sense. No it does not hurt at all just wanted to be sure I understood.

gianfra-t commented on nabla-indexer/test/indexer_tests.ts:
```diff
@@ -0,0 +1,1067 @@
+        (routers) =>
+          routers
+            .find((router) => router.id === address(instance.router))
+            ?.swapPools.find((swapPool) => swapPool.id === address(replacementPool)) !== undefined,
```
Should we also check that the replaced `swapPool` does not exist anymore (not registered with the router)?

TorstenStueber replied:
Yes, I think this is happening here already the `verifyIndexer` that comes up next checks whether there are only two swap pools. Since the first swap pool's reserve changed from `11000000000000` to `20000000000000` and there are only two pools, then this would mean that the indexer does not report the replaced swap pool anymore.

gianfra-t commented on nabla-indexer/test/indexer_tests.ts:
```diff
@@ -0,0 +1,1067 @@
+
+      verifyIndexer(indexerRouters, instance, [
+        {
+          reserve: "12997193504712",
```
Is there an "easy" way to calculate these values without having to use the full equation of the curve? (or at least an approximation to do a sanity check)

TorstenStueber replied:
It is, among others, the root of a quadratic equation. That's why this number looks so random. I think that does not count as "easy" anymore.

gianfra-t commented on nabla-indexer/test/indexer_tests.ts:
```diff
@@ -0,0 +1,1067 @@
+        },
+      ]);
+    },
+
```
Should we check to see if the balances were modified to reflect the swap? Or is it too much coupling on the same test.

TorstenStueber replied:
What balances do you refer to? The pool token balance(s) and/or the LP token balance(s)? As these are not reported by the indexer and, e.g., the Nabla UI reads them directly from the smart contract, I don't actually consider this a necessity of this test suite. Is that what you are referring to?

gianfra-t replied:
I think I was referring to the `tester` balance before and after the swap. What I don't remember is why I thought this could be needed.

TorstenStueber replied:
Is it fine if we leave that out?

gianfra-t reviewed (commented):
Looking good :+1:! I only have minor questions, not on the logic of the test itself. That so far seems sound.
I still haven't run the code locally.

TorstenStueber commented:
I added and adapted the tests in order to incorporate the new changes on this branch: https://github.com/0xamberhq/contracts/pull/111

bogdanS98 commented on README.md:
````diff
@@ -5,11 +5,89 @@ Usage for the example project:
+
+Clone the solang repository and build it:
+
+```shell
````
I think that the path to the solang binary used [here](https://github.com/pendulum-chain/wasm-deploy/blob/ad085183d1698ce4ac43aa75a7b9126b1663b01c/src/actions/compileContract.ts#L77) should be absolute.

ebma replied:
But if we make it absolute then every developer definitely needs to touch that path. If we keep it relative we just have to make sure that the directory is put into the conventional place.

TorstenStueber commented:
I had to add some changes to make the indexer tests run again: this requires an older version of our wrapper contracts because foucoco-standalone is currently incompatible with [the latest main branch of the wrapper contracts](https://github.com/pendulum-chain/pendulum-solidity-wrapper/pull/43) (they now assume that there are multiple chain extensions in place – [as is the case for our our normal runtimes](https://github.com/pendulum-chain/pendulum/pull/437)).

[I changed the dependency](https://github.com/pendulum-chain/wasm-deploy/pull/40/commits/0d703bb860b8ebb8ec600588c4fcfcea2d11d960#diff-357d038e83ad0a421f2dd312fb724851ace406878baf0cab192c49bac04e5264R66) on the wrapper contracts to refer to an older commit.

Furthermore, the tests require a version of foucoco-standalone that includes PR https://github.com/pendulum-chain/foucoco-standalone/pull/10.

ebma commented:
@TorstenStueber I wanted to test this again, this time using our custom Foucoco version that we added to the pendulum repo. That runtime uses the latest chain extensions and thus has to be called with the chain extension ID + function ID as is done by the latest Pendulum solidity wrapper contracts. So I run the Foucoco testchain with
```
cargo run -p pendulum-node -- --chain foucoco-standalone --instant-seal
```

Then I spin up the indexer with
```
sqd process:local
sqd serve
```

but when I do ` npm run test:nabla:local` I get the following error:
```

> wasm-deploy@1.0.0 test:nabla:local
> tsx src/index.ts test nabla --network local

Load project in folder "nabla"
Connecting to chain ws://127.0.0.1:9944
Connected to chain Foucoco-Standalone
Process test suite TestableERC20Tests.ts
Deploying new contract TestableERC20Wrapper
Successfully deployed contract TestableERC20Wrapper to 6h14TV9JVsMwgsgZdLpuEtTYb3kASZDWiwqXJ1Jw9zQq3A9m
Deploying new contract TestableERC20Wrapper
Successfully deployed contract TestableERC20Wrapper to 6mvu7FA1d8bpibtsDxD9junBx2gyEhZfWHzSDfL7QusRjD17


Run test testMintsNative
Deploying new contract TestableERC20Wrapper
Successfully deployed contract TestableERC20Wrapper to 6jCRE1pB1waV1Wdut4F9awnRDSWw5WuZUnUGa1JreYsuk6jK
Deploying new contract TestableERC20Wrapper
Successfully deployed contract TestableERC20Wrapper to 6kN3djWWpyjjcz5gM24KCQfcexRuHKPgxBkis24QKF9s44vn
Run setup code testMintsNative
Run test function testMintsNative
Call contract TestableERC20Wrapper at address 6jCRE1pB1waV1Wdut4F9awnRDSWw5WuZUnUGa1JreYsuk6jK totalSupply []
Done TestableERC20Wrapper totalSupply 199,999,239,999,750,253,600
Call contract TestableERC20Wrapper at address 6jCRE1pB1waV1Wdut4F9awnRDSWw5WuZUnUGa1JreYsuk6jK mint [
  '6k6gXPB9idebCxqSJuqpjPaqfYLQbdLHhvsANH8Dg8GQN3tT',
  10000000000000000n
]

An error occurred
Other
CallError: Other
    at Object.deployedContract.<computed> (/Users/marcel/Documents/wasm-deploy/src/commands/test.ts:253:21)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at Object.testMintsNative (/Users/marcel/Documents/wasm-deploy/nabla/test/TestableERC20Tests.ts:25:7)
    at processTestScripts (/Users/marcel/Documents/wasm-deploy/src/commands/test.ts:417:11)
    at <anonymous> (/Users/marcel/Documents/wasm-deploy/src/commands/test.ts:35:5)
    at createAnimatedTextContext (/Users/marcel/Documents/wasm-deploy/src/utils/terminal.ts:155:5)
    at runTestSuits (/Users/marcel/Documents/wasm-deploy/src/commands/test.ts:34:22)
    at Command.<anonymous> (/Users/marcel/Documents/wasm-deploy/src/commandLine.ts:37:7)
```

and when I do `` I get
```
[...long list of logs of deployed contracts with events...]
Contract event 6jXDP7ZnQXfkgf8634fx5zUFpgJJrqdVD3rRDtWLRK1Czu1U 008eaf04151687736326c9fea17e25fc5287613693c912909cb226aa4794f26a481b791897e40faa497cc132bcd10160bc2f5e03a5a85b2323573df8799efc066f00a0724e18090000000000000000000000000000000000000000000000000000
Decoded event Transfer [
  {
    name: 'from',
    value: '6k6gXPB9idebCxqSJuqpjPaqfYLQbdLHhvsANH8Dg8GQN3tT'
  },
  {
    name: 'to',
    value: '6hVczyTideD7rqQWdzgfMLUxJTrzoqRNEz9ZipWzAjxkw1MW'
  },
  { name: 'value', value: '10,000,000,000,000' }
]
Contract event 6hVczyTideD7rqQWdzgfMLUxJTrzoqRNEz9ZipWzAjxkw1MW 078eaf04151687736326c9fea17e25fc5287613693c912909cb226aa4794f26a4800a0724e1809000000000000000000000000000000000000000000000000000000a0724e18090000000000000000000000000000000000000000000000000000
Decoded event Mint [
  {
    name: 'sender',
    value: '6k6gXPB9idebCxqSJuqpjPaqfYLQbdLHhvsANH8Dg8GQN3tT'
  },
  { name: 'poolSharesMinted', value: '10,000,000,000,000' },
  { name: 'amountPrincipleDeposited', value: '10,000,000,000,000' }
]
Done BackstopPool deposit [ '10,000,000,000,000', '0' ]
stopPrank

An error occurred
(expected true, got false)
AssertionError [AssertError]: (expected true, got false)
    at assertTrue (/Users/marcel/Documents/wasm-deploy/src/testing/stdLib.ts:24:11)
    at Object.testStandard (/Users/marcel/Documents/wasm-deploy/nabla-indexer/test/indexer_tests.ts:273:7)
    at processTestScripts (/Users/marcel/Documents/wasm-deploy/src/commands/test.ts:417:11)
    at <anonymous> (/Users/marcel/Documents/wasm-deploy/src/commands/test.ts:35:5)
    at createAnimatedTextContext (/Users/marcel/Documents/wasm-deploy/src/utils/terminal.ts:155:5)
    at runTestSuits (/Users/marcel/Documents/wasm-deploy/src/commands/test.ts:34:22)
    at Command.<anonymous> (/Users/marcel/Documents/wasm-deploy/src/commandLine.ts:37:7)
```


Do we need to touch the tests again or am I doing something wrong?

TorstenStueber commented:
@ebma Yes, Foucoco Standalone is not working anymore with the `master` branch of the wrapper contracts. I will change the code so that it will use an older commit of the wrapper contracts where the change to multiple chain extensions did not happen.

This is meant to be temporary. Either we change foucoco-standalone to also use multiple chain extensions or we wait for everything [in this comment](https://github.com/pendulum-chain/tasks/issues/297#issuecomment-2125654398) being completed. I prefer the latter.

TorstenStueber commented:
I read your comment more thoroughly now. Your comment was about the test run of the `nabla` project as well as about the `nabla-indexer` project, correct?

For the `nabla-indexer` the problem seems to be that events are not correctly decoded anymore. The auto generated [`decodeEvent` function](https://github.com/pendulum-chain/pendulum-squids/blob/2b6cd874fdc1bfe8552dd829c8d165ef24342bb8/src/abi/backstop.ts#L1706-L1708) is returning events (correctly decoded) but all of them lack the `__kind` field (e.g. [here](https://github.com/pendulum-chain/pendulum-squids/blob/2b6cd874fdc1bfe8552dd829c8d165ef24342bb8/src/abi/backstop.ts#L1958)) that we need to use to distinguish events. For that reason none of the code for event processing is executed anymore.

TorstenStueber commented:
I switched to an older branch where this was working correctly and then back to `main` and it is working correctly there as well now. Strange – I didn't change anything apart from switching branches away from `main` and then back to `main`. But the tests are running correctly now. @ebma it looks like you encountered the same problem here. Not sure what to do about this now. Do you want to test again?

ebma commented:
@TorstenStueber but you investigated this problem already and identified that this is due to some breaking changes in one of the dependencies, right? How should we proceed with this PR?

TorstenStueber commented:
@ebma I waited for https://github.com/pendulum-chain/wasm-deploy/pull/50 to be resolved next will check how to get this PR here back on track.

TorstenStueber committed:
Add local configuration

TorstenStueber committed:
Update tsx dependency

TorstenStueber commented:
@pendulum-chain/devs I added a lot of updates to this PR and updated the description of this PR. Additionally I rebased it so that it is not in conflict with master anymore.

Since this PR has been open and had many changes over many months, I recommend the following course of actions:
* merge this PR
* address the problems raised [in this comment](https://github.com/pendulum-chain/wasm-deploy/pull/40#issuecomment-2166415221) in [this subsequent ticket](https://github.com/pendulum-chain/wasm-deploy/issues/52) instead

gianfra-t commented:
I approved so we can continue with that plan. I believe we also ice boxed the standalone issue right?

TorstenStueber commented:
I would need to check in the next weeks what the latest version in our discussion about the standalone issue was and whether there is any chance to continue. I don't consider this high priority right now as long as we don't change the Nabla code.
